README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# peathamstr

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

‘peathamstr’ extents the
[‘hamstr’](https://github.com/EarthSystemDiagnostics/hamstr) package
(Hierarchical Accumulation Modelling with Stan and R) (Dolman 2023) with
a modified version of Clymo’s Bog Growth Model (Clymo 1984).

The extension uses ‘hamstr’ to compute an age-depth model, but also
estimates layer-specific peat addition rates to the catotelm (PAR) and a
global exponential decomposition rate. In addition, peat mass fluxes net
carbon uptake (NCU) and net carbon release (NCR) as defined in Yu (2011)
are computed.

This allows to estimate catotelm PAR and decomposition rates also for
peat cores where the assumption of a constant PAR of Clymo’s Bog Growth
Model (Clymo 1984) is violated, e.g. when PAR declined over time (see
for example Yu et al. (2003)).

The package shamelessly recycles much of the code from the
[‘hamstr’](https://github.com/EarthSystemDiagnostics/hamstr) package and
much credit goes to its developers. When using ‘peathamstr’, do not
forget to also cite the ‘hamstr’ package!

## Installation

You can install the development version of ‘peathamstr’ like so:

``` r
remotes::install_github("henningte/peathamstr")
```

## Example

This example shows how to estimate the decomposition rate, net carbon
uptake (NCU) and net carbon release (NCR) for a peat core with
non-constant peat addition rate (PAR).

``` r
library(peathamstr)
library(rstan)
#> Loading required package: StanHeaders
#> 
#> rstan version 2.26.22 (Stan version 2.26.1)
#> For execution on a local, multicore CPU with excess RAM we recommend calling
#> options(mc.cores = parallel::detectCores()).
#> To avoid recompilation of unchanged Stan programs, we recommend calling
#> rstan_options(auto_write = TRUE)
#> For within-chain threading using `reduce_sum()` or `map_rect()` Stan functions,
#> change `threads_per_chain` option:
#> rstan_options(threads_per_chain = 1)
#> Do not specify '-march=native' in 'LOCAL_CPPFLAGS' or a Makevars file
library(magrittr)
#> 
#> Attaching package: 'magrittr'
#> The following object is masked from 'package:rstan':
#> 
#>     extract
library(pangaear) # to download example data
#> Registered S3 method overwritten by 'httr':
#>   method           from  
#>   print.cache_info hoardr
library(ggplot2)
```

Download and reshape data (Yu 2018b, 2018a) from the Pangaea database:

``` r
d1 <- pangaear::pg_data(doi = '10.1594/PANGAEA.890357')
#> Downloading 1 datasets from 10.1594/PANGAEA.890357
#> Processing 1 files
d2 <- pangaear::pg_data(doi = '10.1594/PANGAEA.890392')
#> Downloading 1 datasets from 10.1594/PANGAEA.890392
#> Processing 1 files

d <- 
  dplyr::full_join(
    d1[[1]]$data |>
      dplyr::select(1, 3, 6) |>
      setNames(nm = c("depth_midpoint", "bulk_density", "C")) |>
      dplyr::filter(!is.na(bulk_density)) |>
      dplyr::mutate(
        C = C/bulk_density,
        depth_midpoint = round(depth_midpoint * 100, 1), # assumed
        thickness = 1, # assumed
        mass = bulk_density * thickness * 10000/1000,
        cumulative_mass = c(cumsum(mass))
      ),
    d2[[1]]$data |>
      dplyr::select(2, 4, 5) |>
      setNames(nm = c("depth_midpoint", "age", "age_sd")) |>
      dplyr::filter(!is.na(age)) |>
      dplyr::mutate(
        depth_midpoint = round(depth_midpoint * 100, 1),
        age = age * 1000,
        age_sd = age_sd * 1000
      ) |>
      hamstr::calibrate_14C_age(age.14C = "age", age.14C.se = "age_sd") |>
      dplyr::select(-age, -age_sd) |>
      dplyr::rename(
        age = "age.14C.cal",
        age_sd = "age.14C.cal.se"
      ),
    by = "depth_midpoint"
  ) |>
  dplyr::arrange(depth_midpoint) |>
  dplyr::filter(depth_midpoint <= max(depth_midpoint[!is.na(age)]) & depth_midpoint >= 0) |>
  dplyr::mutate(
    depth_upper = depth_midpoint - 0.5,
    depth_lower = depth_midpoint + 0.5
  )

# set.seed(23)
# d <- 
#   dplyr::bind_rows(
#     d |>
#       dplyr::filter(! is.na(age)),
#     d |>
#       dplyr::filter(is.na(age)) |>
#       dplyr::slice_sample(n = floor(nrow(d)/3), replace = FALSE)
#   ) |>
#   dplyr::arrange(depth_upper)

# summarize the data to make the example faster to compute
# target_depths <- 
#   tibble::tibble(
#     sample_depth_upper = c(seq(min(d$depth_upper), 60, by = 2), seq(60, max(d$depth_lower), by = 20)),
#     sample_depth_lower = c(sample_depth_upper[-1], max(d$depth_lower)) 
#   )

# d <- 
#   dplyr::bind_rows(
#     d |>
#       dplyr::filter(!is.na(age)),
#     d |>
#       dplyr::filter(is.na(age)) %>%
#       dplyr::mutate(
#         index_target_depth =
#           purrr::map_int(seq_len(nrow(.)), function(i) {
#             index <- depth_lower[[i]] <= target_depths$sample_depth_lower & depth_upper[[i]] >= target_depths$sample_depth_upper
#             if(sum(index) == 0) {
#               NA_integer_
#             } else {
#               which(index)[[1]]
#             }
#           })
#       ) |>
#       dplyr::group_by(index_target_depth) |>
#       dplyr::summarise(
#         dplyr::across(!dplyr::all_of(c("depth_upper", "depth_lower", "depth_midpoint", "cumulative_mass", "mass", "thickness")), mean),
#         depth_upper = min(depth_upper),
#         depth_lower = max(depth_lower),
#         mass = sum(mass),
#         cumulative_mass = max(cumulative_mass),
#         .groups = "drop"
#       )
#   ) |>
#   dplyr::mutate(
#     thickness = depth_lower - depth_upper,
#     depth_midpoint = depth_upper + (thickness) * 0.5
#   ) |>
#   dplyr::select(-index_target_depth) |>
#   dplyr::arrange(depth_upper)
```

Show the cumulative mass-age curve (here using only point age values and
cumulative masses for dated layers). It is clearly visible that the core
violates the assumptions of the original Bog Growth Model (see also Yu
et al. (2003) for a detailed description of the peatland).

``` r
d |>
  dplyr::filter(!is.na(age)) |>
  ggplot(aes(y = cumsum(mass * C), x = age)) +
  geom_path() +
  geom_point() +
  labs(y = expression("Cumulative carbon mass (kg m"^{-2}*")"), x = "age (yr BP)")
```

<img src="man/figures/README-example-3-1.png" width="60%" />

Estimate the posterior distribution (computes the age-depth model and
fits the modified Clymo model with PAR varying over time).

``` r
fit_1 <- 
  peat_hamstr(
    K = c(3, 3, 3),
    depth = 
      d |> 
      dplyr::filter(!is.na(age)) |> 
      dplyr::pull(depth_midpoint),
    obs_age =
      d |> 
      dplyr::filter(!is.na(age)) |> 
      dplyr::pull(age),
    obs_err = 
      d |> 
      dplyr::filter(!is.na(age)) |> 
      dplyr::pull(age_sd),
    layer_mass = 
      d |> 
      dplyr::filter(is.na(age)) |>
      dplyr::filter(!is.na(mass)) |> 
      dplyr::pull(cumulative_mass),
    cumulative_mass0 = d$cumulative_mass[[1]] - d$mass[[1]],
    # depth2 =
      # d |> 
      # dplyr::filter(!is.na(cumulative_mass)) |> 
      # dplyr::pull(depth_midpoint),
    depth_clymo_par_constant = 10,
    depth2_upper =
      d |> 
      dplyr::filter(is.na(age)) |>
      dplyr::filter(!is.na(cumulative_mass)) |> 
      dplyr::pull(depth_upper),
    depth2_lower =
      d |> 
      dplyr::filter(is.na(age)) |>
      dplyr::filter(!is.na(cumulative_mass)) |> 
      dplyr::pull(depth_lower),
    min_age = 0,
    top_depth = 0,
    bottom_depth = max(d$depth_lower),
    p1_clymo_par = 8,
    p2_clymo_par = 8/0.2,
    p3_clymo_par = 1/20,
    clymo_par_memory_p1 = 2,
    clymo_par_memory_p2 = 2,
    p1_clymo_alpha_1 = 10, 
    p2_clymo_alpha_1 = 10/0.04, 
    p3_clymo_alpha_1 = 1/100,
    p1_age0 = 20,
    p2_age0 = 20/0.2,
    p3_age0 = 1/20,
    p1_ac_age = 20,
    p2_ac_age = 20/200,
    p3_ac_age = 40,
    # the seed argument for the sampler is set here so that
    # this example always returns the same numerical result
    stan_sampler_args = 
      list(
        seed = 34564, 
        chains = 4, 
        iter = 4000, 
        cores = 4, 
        control = list(max_treedepth = 15)
      )
  )
#> Warning: The largest R-hat is 2.03, indicating chains have not mixed.
#> Running the chains for more iterations may help. See
#> https://mc-stan.org/misc/warnings.html#r-hat
#> Warning: Bulk Effective Samples Size (ESS) is too low, indicating posterior means and medians may be unreliable.
#> Running the chains for more iterations may help. See
#> https://mc-stan.org/misc/warnings.html#bulk-ess
#> Warning: Tail Effective Samples Size (ESS) is too low, indicating posterior variances and tail quantiles may be unreliable.
#> Running the chains for more iterations may help. See
#> https://mc-stan.org/misc/warnings.html#tail-ess
```

Show the age-depth model (this is the same function as in the original
‘hamstr’ package):

``` r
plot(fit_1, type = "default")
```

<img src="man/figures/README-example-5-1.png" width="100%" />

Show estimated and measured cumulative masses versus depth:

``` r
# plot(fit_1, type = "cumulative_mass_profile")
plot(
  fit_1, 
  type = "cumulative_mass_profile"
)
#> Warning: Returning more (or less) than 1 row per `summarise()` group was deprecated in
#> dplyr 1.1.0.
#> ℹ Please use `reframe()` instead.
#> ℹ When switching from `summarise()` to `reframe()`, remember that `reframe()`
#>   always returns an ungrouped data frame and adjust accordingly.
#> ℹ The deprecated feature was likely used in the peathamstr package.
#>   Please report the issue at <https://github.com/henningte/peathamstr/issues>.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.
```

<img src="man/figures/README-example-6-1.png" width="60%" style="display: block; margin: auto;" />

Plot a histogram of the estimate exponential decomposition rates for the
acrotelm and catotelm and of the acrotelm-catotelm boundary age:

``` r
as.data.frame(fit_1$fit, pars = c("clymo_alpha_1", "clymo_alpha_2", "ac_age")) |>
  tidyr::pivot_longer(
    cols = dplyr::everything(),
    names_to = "variable",
    values_to = "value"
  ) |>
  ggplot(aes(x = value)) +
  geom_histogram(bins = 30) +
  labs(y = "Count", x = expression(alpha~"("*yr^{-1}*") or age (yr)")) +
  facet_wrap(~ variable, scales = "free_x")
```

<img src="man/figures/README-example-7-1.png" width="100%" style="display: block; margin: auto;" />

Plot for net mass uptake (NMU) and net mass release (NMR) can also be
created:

``` r
plot(fit_1, type = "mass_fluxes") + 
  facet_wrap(~ variable, ncol = 1L, scales = "free_y")
```

<img src="man/figures/README-example-10-1.png" width="60%" style="display: block; margin: auto;" />

# References

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-Clymo.1984" class="csl-entry">

Clymo, R S. 1984. “The Limits to Peat Bog Growth.” *Philosophical
Transactions of the Royal Society of London. Series B, Biological
Sciences* 303 (1117): 51.

</div>

<div id="ref-Dolman.2023" class="csl-entry">

Dolman, Andrew. 2023. “<span class="nocase">hamstr</span>: Hierarchical
Accumulation Modelling with Stan and R.”

</div>

<div id="ref-Yu.2011" class="csl-entry">

Yu, Zicheng. 2011. “Holocene Carbon Flux Histories of the World’s
Peatlands: Global Carbon-Cycle Implications.” *The Holocene* 21 (5):
761–74. <https://doi.org/10.1177/0959683610386982>.

</div>

<div id="ref-Yu.2018a" class="csl-entry">

———. 2018a. “Age Determination of Upper\_Pinto Peat Core.” PANGAEA.
<https://doi.org/10.1594/PANGAEA.890392>.

</div>

<div id="ref-Yu.2018" class="csl-entry">

———. 2018b. “Geochemistry of Upper\_Pinto Peat Core.” PANGAEA.
<https://doi.org/10.1594/PANGAEA.890357>.

</div>

<div id="ref-Yu.2003" class="csl-entry">

Yu, Zicheng, Dale H Vitt, Ian D Campbell, and Michael J Apps. 2003.
“Understanding Holocene Peat Accumulation Pattern of Continental Fens in
Western Canada.” *Can. J. Bot.* 81: 16.

</div>

</div>


NEWS.md
# peathamstr 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.

## New models

* Add a new model version which estimates a separate decomposition rate for the acrotelm and catotelm and the time a newly formed layer spends in the acrotelm.

## Changes
* Adapt the functions to this new model.

## Bug fixes
* Remove the distinction between C mass fluxes and mass fluxes because this makes no sense: If we want C fluxes, we need to compute the model with C masses.


R/peat_hamstr.R
#' Estimates an age-depth model and parameters from a modified Clymo Bog Growth Model
#'
#' @inheritParams hamstr::hamstr
#'
#' @param cumulative_mass A numeric vector with the cumulative mass of the peat
#' layers (kg m\eqn{^{-1}}).
#'
#' @param cumulative_mass0 A numeric value with the cumulative mass of all peat
#' layers above the first considered peat layer (kg m\eqn{^{-1}}).
#'
#' @param depth2 A numeric vector with the same length as `cumulative_mass`
#' defining the depth of the corresponding peat layers.
#'
#' @param depth2_upper,depth2_lower A numeric vector with the same length as
#' `cumulative_mass` defining the depth of the upper and lower boundaries of the
#' corresponding peat layers.
#'
#' @param p1_layer_mass_shape,p2_layer_mass_shape A numeric
#' value representing the shape and rate parameter of a Gamma distribution
#' modeling the shape parameter of the Gamma distribution assumed for modeled
#' peat cumulative masses.
#'
#' @param p1_clymo_par,p2_clymo_par A numeric
#' value representing the shape and rate parameter of the Gamma distribution
#' modeling peat mass addition rates (g m\eqn{^{-2}} yr\eqn{^{-1}}).
#'
#' @param p1_clymo_alpha,p2_clymo_alpha A numeric
#' value representing the shape and rate parameter of the Gamma distribution
#' modeling peat decomposition rate (in a simple exponential decomposition
#' model) (yr\eqn{^{-1}}).
#'
#' @export
peat_hamstr <-
  function(depth, obs_age, obs_err,
           min_age = 1950 - as.numeric(format(Sys.Date(), "%Y")),
           K = NULL,
           top_depth = NULL, bottom_depth = NULL,
           acc_mean_prior = NULL,
           acc_shape = 1.5,
           mem_mean = 0.5, mem_strength = 10,
           model_bioturbation = FALSE,
           n_ind = NULL,
           L_prior_mean = 10,
           L_prior_shape = 2,
           L_prior_sigma = NULL,
           model_displacement = FALSE,
           D_prior_scale = 10,
           model_hiatus = FALSE,
           H_top = NULL, H_bottom = NULL,
           layer_mass,
           cumulative_mass0,
           depth_clymo_par_constant,
           depth2_upper,
           depth2_lower,
           p1_layer_mass_shape = 5,
           p2_layer_mass_shape = 0.01,
           p3_layer_mass_shape = 100,
           p1_clymo_par = 3,
           p2_clymo_par = 0.02,
           p3_clymo_par = 100,
           p1_clymo_alpha_1 = 2,
           p2_clymo_alpha_1 = 50,
           p3_clymo_alpha_1 = 1/1000,
           p1_clymo_alpha_2 = 3,
           p2_clymo_alpha_2 = 3/0.005,
           p3_clymo_alpha_2 = 1/10000,
           p1_ac_age = 3,
           p2_ac_age = 3/80,
           p3_ac_age = 40,
           clymo_par_memory_p1 = 1,
           clymo_par_memory_p2 = 1,
           p1_age0 = 5,
           p2_age0 = 5/1,
           p3_age0 = 1,
           sample_posterior = TRUE,
           hamstr_control = list(),
           stan_sampler_args = list()
  ) {

    # # first: get hamstr_fit object
    # res <-
    #   hamstr::hamstr(
    #     depth = depth,
    #     obs_age = obs_age,
    #     obs_err = obs_err,
    #     min_age = min_age,
    #     K = K,
    #     top_depth = top_depth,
    #     bottom_depth = bottom_depth,
    #     acc_mean_prior = acc_mean_prior,
    #     acc_shape = acc_shape,
    #     mem_mean = mem_mean,
    #     mem_strength = mem_strength,
    #     model_bioturbation = model_bioturbation,
    #     n_ind = n_ind,
    #     L_prior_mean = L_prior_mean,
    #     L_prior_shape = L_prior_shape,
    #     L_prior_sigma = L_prior_sigma,
    #     model_displacement = model_displacement,
    #     D_prior_scale = D_prior_scale,
    #     model_hiatus = model_hiatus,
    #     H_top = H_top,
    #     H_bottom = H_bottom,
    #     sample_posterior = FALSE,
    #     hamstr_control = hamstr_control,
    #     stan_sampler_args = stan_sampler_args
    #   )

    # ## next: add additional parts

    # # data
    # res$data$cumulative_mass <- cumulative_mass
    # res$data$cumulative_mass0 <- cumulative_mass0
    # res$data$p1_layer_mass_shape <- p1_layer_mass_shape
    # res$data$p2_layer_mass_shape <- p2_layer_mass_shape
    # res$data$p1_clymo_par <- p1_clymo_par
    # res$data$p2_clymo_par <- p2_clymo_par
    # res$data$p1_clymo_alpha <- p1_clymo_alpha
    # res$data$p1_clymo_alpha <- p2_clymo_alpha
    # res$data$depth2 <- depth2
    # res$data$N2 <- length(res$data$depth2)
    # res$data$which_c2 <-
    #   sapply(
    #     res$data$depth2,
    #     function(d) {
    #       which.max((res$data$c_depth_bottom < d) * (res$data$c_depth_bottom - d))
    #     })

    # used_sampler_args <- do.call(hamstr:::get_stan_sampler_args, stan_sampler_args)


    # # set the seed here and not inside get_inits_hamstr, so that the chains are
    # # different

    # set.seed(used_sampler_args$seed)

    # inits <- replicate(used_sampler_args$chains, #---todo: avoid using internal functions
    #                    list(hamstr:::get_inits_hamstr(res$data)))

    # args <- #---todo
    #   list(
    #     object = stanmodels$peathamstr,
    #     data = res$data,
    #     init = inits
    #   )

    # args <- append(args, used_sampler_args)

    # if(sample_posterior) {
    #   res$fit <- do.call(rstan::sampling, args)
    # }


    ###

    stan_dat <- peat_make_stan_dat_hamstr()
    used_sampler_args <- do.call(hamstr:::get_stan_sampler_args, stan_sampler_args)
    set.seed(used_sampler_args$seed)
    inits <- replicate(used_sampler_args$chains, list(hamstr:::get_inits_hamstr(stan_dat)))
    args <- list(object = stanmodels$mm2, data = stan_dat,
                 init = inits)
    args <- append(args, used_sampler_args)
    if (sample_posterior) {
      fit <- do.call(rstan::sampling, args)
    }
    else if (sample_posterior == FALSE) {
      fit <- NA
    }
    stan_dat <- append(stan_dat, used_sampler_args)
    info <- list(version = utils::packageVersion("hamstr"), time = Sys.time())
    out <- list(fit = fit, data = stan_dat, info = info)
    class(out) <- append(c("peathamstr_fit", "hamstr_fit"), class(out))
    return(out)

  }


peat_make_stan_dat_hamstr <- function (...)
{
  l <- c(as.list(parent.frame()))
  default.args <- formals(peat_hamstr)
  default.args$cumulative_mass <- as.name("layer_mass")
  default.args$cumulative_mass0 <- as.name("cumulative_mass0")
  default.args$p1_layer_mass_shape <- as.name("p1_layer_mass_shape")
  default.args$p2_layer_mass_shape <- as.name("p2_layer_mass_shape")
  default.args$p3_layer_mass_shape <- as.name("p3_layer_mass_shape")
  default.args$p1_clymo_par <- as.name("p1_clymo_par")
  default.args$p2_clymo_par <- as.name("p2_clymo_par")
  default.args$p3_clymo_par <- as.name("p3_clymo_par")
  default.args$p1_clymo_alpha_1 <- as.name("p1_clymo_alpha_1")
  default.args$p2_clymo_alpha_1 <- as.name("p2_clymo_alpha_1")
  default.args$p3_clymo_alpha_1 <- as.name("p3_clymo_alpha_1")
  default.args$p1_clymo_alpha_2 <- as.name("p1_clymo_alpha_2")
  default.args$p2_clymo_alpha_2 <- as.name("p2_clymo_alpha_2")
  default.args$p3_clymo_alpha_2 <- as.name("p3_clymo_alpha_2")
  default.args$p1_ac_age <- as.name("p1_ac_age")
  default.args$p2_ac_age <- as.name("p2_ac_age")
  default.args$p3_ac_age <- as.name("p3_ac_age")
  default.args$depth2_upper <- as.name("depth2_upper")
  default.args$depth2_lower <- as.name("depth2_lower")
  default.args$clymo_par_memory_p1 <- as.name("clymo_par_memory_p1")
  default.args$clymo_par_memory_p2 <- as.name("clymo_par_memory_p2")
  default.args$clymo_par_memory_p1 <- as.name("clymo_par_memory_p1")
  default.args$p1_age0 <- as.name("p1_age0")
  default.args$p2_age0 <- as.name("p2_age0")
  default.args$p3_age0 <- as.name("p3_age0")
  default.arg.nms <- names(default.args)
  l <- l[lapply(l, is.null) == FALSE]
  default.args[names(l)] <- l
  l <- default.args
  hc.default.args <- formals(hamstr:::hamstr_control)
  hc.default.arg.nms <- names(hc.default.args)
  hc <- l$hamstr_control
  hc <- hc[lapply(hc, is.null) == FALSE]
  hc.default.args[names(hc)] <- hc
  hc <- hc.default.args
  l <- append(l, hc)
  l <- l[names(l) != "hamstr.control"]
  if (is.null(l$acc_mean_prior)) {
    d <- data.frame(depth = l$depth, obs_age = l$obs_age)
    acc_mean <- stats::coef(MASS::rlm(obs_age ~ depth, data = d))[2]
    acc_mean <- signif(acc_mean, 2)
    if (acc_mean <= 0) {
      warning("Estimated mean accumulation rate is negative - using value = 20")
      acc_mean <- 20
    }
    l$acc_mean_prior <- acc_mean
  }
  ord <- order(l$depth)
  l$depth <- l$depth[ord]
  l$obs_age <- l$obs_age[ord]
  l$obs_err <- l$obs_err[ord]
  if (l$model_bioturbation == TRUE) {
    if (is.null(l$L_prior_sigma) == FALSE)
      message("L_prior_shape is being overriden by L_prior_sigma.")
    if (length(c(l$L_prior_sigma, l$L_prior_shape)) == 0)
      stop("One of either L_prior_sigma or L_prior_shape must be specified.\n             Set either to 0 to impose a fixed mixing depth.")
    if ((length(l$n_ind) == 1 | length(l$n_ind) == length(l$obs_age)) ==
        FALSE)
      stop("n_ind must be either a single value or a vector the same length as obs_age")
    if (length(l$n_ind == 1))
      l$n_ind <- rep(l$n_ind, length(l$obs_age))
    l$n_ind <- l$n_ind[ord]
    if (is.null(l$L_prior_sigma) == FALSE) {
      if (l$L_prior_sigma == 0)
        l$L_prior_shape <- 0
      else l$L_prior_shape <- gamma_sigma_shape(mean = l$L_prior_mean,
                                                sigma = l$L_prior_sigma)$shape
    }
  }
  else if (l$model_bioturbation == FALSE) {
    l$n_ind <- numeric(0)
  }
  if (is.null(l$infl_sigma_sd)) {
    l$infl_sigma_sd <- 10 * mean(l$obs_err)
  }
  if (l$min_age > min(l$obs_age)) {
    warning("min_age is older than minimum obs_age")
  }
  if (is.null(l$top_depth))
    l$top_depth <- l$depth[1]
  if (is.null(l$bottom_depth))
    l$bottom_depth <- utils::tail(l$depth, 1)
  depth_range <- l$bottom_depth - l$top_depth
  if (l$top_depth > min(l$depth))
    stop("top_depth must be above or equal to the shallowest data point")
  if (l$bottom_depth < max(l$depth))
    stop("bottom_depth must be deeper or equal to the deepest data point")
  if (is.null(l$K)) {
    K_fine_1 <- l$bottom_depth - l$top_depth
    min.d.depth <- stats::median(diff(sort(unique(l$depth))))
    K_fine_2 <- round(16 * K_fine_1/min.d.depth)
    K_fine <- min(c(K_fine_1, K_fine_2))
    if (K_fine > 900)
      K_fine <- 900
    l$K <- hamstr:::default_K(K_fine)
  }
  l$N <- length(l$depth)
  stopifnot(l$N == length(l$obs_err), l$N == length(l$obs_age))
  alpha_idx <- hamstr:::alpha_indices(l$K)
  l$K_tot <- sum(alpha_idx$nK)
  l$K_fine <- utils::tail(alpha_idx$nK, 1)
  l$c <- 1:l$K_fine
  l$mem_alpha = l$mem_strength * l$mem_mean
  l$mem_beta = l$mem_strength * (1 - l$mem_mean)
  l$mem_mean = l$mem_mean
  l$mem_strength = l$mem_strength
  l$delta_c = depth_range/l$K_fine
  l$c_depth_bottom = l$delta_c * l$c + l$top_depth
  l$c_depth_top = c(l$top_depth, l$c_depth_bottom[1:(l$K_fine - 1)])
  l$modelled_depths <- c(l$c_depth_top[1], l$c_depth_bottom)
  l$which_c = sapply(l$depth, function(d) which.max((l$c_depth_bottom < d) * (l$c_depth_bottom - d)))
  l <- append(l, alpha_idx)
  l$n_lvls <- length(l$K)
  l$scale_shape = as.numeric(l$scale_shape)
  l$model_bioturbation = as.numeric(l$model_bioturbation)
  l$model_displacement = as.numeric(l$model_displacement)
  l$smooth_s = as.numeric(l$smooth_s)
  l$model_hiatus = as.numeric(l$model_hiatus)
  if (is.null(l$H_top))
    l$H_top = l$top_depth
  if (is.null(l$H_bottom))
    l$H_bottom = l$bottom_depth
  l$smooth_i <- hamstr:::get_smooth_i(l, l$L_prior_mean)
  l$I <- nrow(l$smooth_i)

  ## clymo model

  # first: construct sections for mass data
  c2 <-
    tibble::tibble(
      sample_depth_upper = l$depth2_upper,
      sample_depth_lower = l$depth2_lower,
      index_has_mass_measurements = TRUE
    )

  # add missing sections to make the sequence contiguous
  c2 <-
    dplyr::bind_rows(
      c2,
      purrr::map_dfr(seq_len(nrow(c2))[-1], function(i) {
        tibble::tibble(
          sample_depth_upper = c2$sample_depth_lower[i - 1],
          sample_depth_lower = c2$sample_depth_upper[i],
          index_has_mass_measurements = FALSE
        )
      })
    ) |>
    dplyr::mutate(
      sample_thickness = sample_depth_lower - sample_depth_upper,
      sample_depth_mean = sample_depth_upper + sample_thickness * 0.5
    ) |>
    dplyr::filter(sample_thickness != 0.0) |>
    dplyr::arrange(sample_depth_upper)

  l$index_clymo_par_constant <- which(c2$sample_depth_lower >= l$depth_clymo_par_constant & c2$sample_depth_upper <= l$depth_clymo_par_constant)[[1]]
  l$N2 <- sum(c2$index_has_mass_measurements)
  l$N2_c <- nrow(c2)
  l$index_has_mass_measurements <- which(c2$index_has_mass_measurements)
  l$depth2_all <- c(c2$sample_depth_upper, tail(c2$sample_depth_lower, 1L))
  l$which_c2_all <- sapply(l$depth2_all, function(d) which.max((l$c_depth_bottom < d) * (l$c_depth_bottom - d)))
  l$which_c2_lower <- sapply(c2$sample_depth_lower, function(d) which.max((l$c_depth_bottom < d) * (l$c_depth_bottom - d)))

  return(l)
}


#' Plots objects of class `peathamstr_fit`
#'
#' @param x An object of class `peathamstr_fit`.
#'
#' @param type A character value. Must be one of the plot `type`s available for
#' `hamstr_fit` objects (see `hamstr::plot.hamstr_fit()`) or the additional types
#' provided by the 'peathamstr' package:
#' \describe{
#'   \item{`"cumulative_mass_profile"`}{Plots the modeled and measured
#'    cumulative masses versus depth.}
#'   \item{`"mass_fluxes"`}{Plots the modeled and measured
#'    net mass uptake (NMU) and net mass release (NMR) versus average estimated
#'    ages. Shaded areas are 50\% and 95\% posterior intervals.}
#' }
#'
#' @param summarise See `hamstr::plot.hamstr_fit()`.
#'
#' @param ... Additional paramters (currently ignored).
#'
#' @return An object of class `ggplot`.
#'
#' @export
plot.peathamstr_fit <- function(
    x,
    type =
      c("default",
        "age_models",
        "acc_rates",
        "hier_acc_rates",
        "acc_mean_prior_post",
        "mem_prior_post",
        "L_prior_post",
        "D_prior_post",
        "PDF_14C",
        "cumulative_mass_profile",
        "mass_fluxes"
      ),
    summarise,
    ...)
{

  if(type %in% c("default", "age_models", "acc_rates", "hier_acc_rates", "acc_mean_prior_post", "mem_prior_post", "L_prior_post", "D_prior_post", "PDF_14C")) {
    NextMethod()
  } else {
    switch(
      type,
      "cumulative_mass_profile" = {

          dplyr::bind_rows(
            tibble::tibble(
              iter = 1L,
              depth_upper = x$data$depth2_upper,
              depth_lower = x$data$depth2_lower,
              depth = x$data$depth2_upper,
              y = x$data$layer_mass,
              variable_type = "observed"
            ),
            predict(
              object = x,
              type = "cumulative_mass",
              depth = "modelled",
              carbon_content = NULL
            ) |>
              dplyr::rename(
                y = "cumulative_mass",
                depth_upper = "depth2_upper",
                depth_lower = "depth2_lower"
              ) |>
              dplyr::mutate(
                variable_type = "predicted"
              )
          ) |>
          dplyr::group_by(
            .data$variable_type, .data$depth, .data$depth_upper, .data$depth_lower
          ) |>
          summarise_q(var = .data$y, probs = c(0.025, 0.159, 0.25, 0.5, 0.75, 0.841, 0.975)) |>
          ggplot2::ggplot(ggplot2::aes(y = .data$mean, x = .data$depth)) +
          ggplot2::geom_errorbar(ggplot2::aes(ymin = .data$`2.5%`, ymax = .data$`97.5%`, group = .data$variable_type), color = "grey50", width = 0) +
          ggplot2::geom_point(ggplot2::aes(fill = .data$variable_type), shape = 21) +
          ggplot2::labs(y = expression("Cumulative mass (kg m"^{-2}*")"), x = "Depth of upper layer boundary (cm)") +
          ggplot2::guides(color = ggplot2::guide_legend(title = ""))

      },
      "mass_fluxes" = {

        summary(
          object = x,
          type = type,
          probs = c(0.025, 0.159, 0.25, 0.5, 0.75, 0.841, 0.975)
        ) |>
          dplyr::mutate(
            variable = toupper(variable)
          ) |>
          ggplot2::ggplot(ggplot2::aes(y = .data$mean, x = .data$age)) +
          geom_hline(yintercept = 0, color = "grey50") +
          ggplot2::geom_ribbon(ggplot2::aes(ymin = .data$`2.5%`, ymax = .data$`97.5%`, fill = .data$variable), color = NA, alpha = 0.3) +
          ggplot2::geom_ribbon(ggplot2::aes(ymin = .data$`25%`, ymax = .data$`75%`, fill = .data$variable), color = NA, alpha = 0.3) +
          ggplot2::geom_path(ggplot2::aes(color = .data$variable)) +
          ggplot2::labs(y = expression("Mass flux (kg m"^{-2}~yr^{-1}*")"), x = "Average estimated age (yr)") +
          ggplot2::guides(fill = ggplot2::guide_legend(title = "Mass flux"), color = ggplot2::guide_legend(title = "Mass flux"))

      })
  }


}
